### PR TITLE
Javascript bank

### PR DESCRIPTION
**Pull request by Quintin-Russell**
image

### CODE REVIEW

File: javascript-bank/account.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    this.number -= amt;
+    return true
+  } else {
+    return false
+  }
+};
+
+Account.prototype.getBalance = function () {
+  return this.balance
+};
```
Comment (TimelikeClosure) [inline javascript-bank/account.js:35]: The balance should be calculated in the `getBalance` method, not beforehand.

---

image

File: javascript-bank/account.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    return false
+  }
+};
+
+Account.prototype.withdraw = function (amt) {
+  if ((amt > 0) && ((amt % 1) === 0)) {
+    let trans = new Transaction('withdrawal', amt);
+    this.transactions.push(trans);
+    this.balance -= amt;
+    this.number -= amt;
```
Comment (TimelikeClosure) [inline javascript-bank/account.js:26]: The `number` property is a unique account number identifier and should not change for the life of the account.

---

image

File: javascript-bank/bank.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
+/* exported Bank */
+function Bank() {
+  this.accounts = [];
+  this.nextAccountNumber = 1;
+}
+
+Bank.prototype.openAccount = function (holder, balance) {
+  if ((balance > 0) && (balance % 1 === 0)) {
+    let acc = new Account(balance, holder);
```
Comment (TimelikeClosure) [inline javascript-bank/bank.js:9]: The first parameter of the `Account` constructor is the account number, not the balance.

---

image

File: javascript-bank/bank.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    acc.nextAccountNumber = this.nextAccountNumber;
+    acc.balance = balance;
```
Comment (TimelikeClosure) [inline javascript-bank/bank.js:11]: In line with "encapsulation", external code should not directly modify an `Account` instance's properties. It should make use of the `Account`'s methods instead.

File: javascript-bank/bank.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    this.nextAccountNumber +=1;
+    return (this.nextAccountNumber - 1)
+  } else {
+    return null
+  }
+};
+
+Bank.prototype.getAccount = function (number) {
+  let account = null;
+  for (let acc of this.accounts) {
```
Comment (TimelikeClosure) [inline javascript-bank/bank.js:22]: Can you explain how `for ... of` loops work? If so, great! If not, use something more basic.

File: javascript-bank/bank.js [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    if (acc.nextAccountNumber === number) {
```
Comment (TimelikeClosure) [inline javascript-bank/bank.js:23]: There should not be a `nextAccountNumber` property on `Account`s.

---

image

File: javascript-bank/bank.js [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+      break
+    };
+  };
+  return account
+};
+
+Bank.prototype.getTotalAssets = function() {
+  let total = 0;
+  for (let acc of this.accounts) {
+    total += acc.balance
```
Comment (TimelikeClosure) [inline javascript-bank/bank.js:34]: There should not be a `balance` property on `Account`s:

---

image

**Review by TimelikeClosure — CHANGES_REQUESTED** [review]

**Review by TimelikeClosure — CHANGES_REQUESTED** [review]
@Quintin-Russell A couple of the previous issues weren't fixed.

**Comment by Quintin-Russell** [comment]
could you please be more specific next time. thank you!

**Review by TimelikeClosure — APPROVED** [review]

**Pull request merged**